File: 2021/Day 2/Javascript/code.js
exports.part1 = (input) => {
    let horizontalPosition = 0;
    let depth = 0;

    let instruction = {
        "forward": (n) => {
            horizontalPosition += n;
        },
        "down": (n) => {
            depth += n;
        },
        "up": (n) => {
            depth -= n;
        }
    }

    input.split("\n").map(c => {
        let [i, v] = c.split(" ");
        v = parseInt(v);
        instruction[i](v);
    });

    return depth * horizontalPosition;
}

exports.part2 = (input) => {
    let horizontalPosition = 0;
    let depth = 0;
    let aim = 0;

    instruction = {
        "forward": (n) => {
            horizontalPosition += n;
            depth += aim * n;
        },
        "down": (n) => {
            aim += n;
        },
        "up": (n) => {
            aim -= n;
        }
    }

    input.split("\n").map(c => {
        let [i, v] = c.split(" ");
        v = parseInt(v);
        instruction[i](v);
    });

    return horizontalPosition * depth;
}

File: 2021/Day 2/lua/code.lua
function dofile(filename) return assert(loadfile(debug.getinfo(2, "S").source:sub(2):match("(.*/)") .. "/" .. filename))() end

local Array = dofile("./../../Common/lua/Array.lua")
local Utility = dofile("./../../Common/lua/Utility.lua")

local code = {}

function code:part1(input)
	local input = Utility.split(input, "\n", true)

	local horizontalPosition = 0
	local depth = 0
	local instruction = {
		forward = function(n)
			horizontalPosition = horizontalPosition + n
		end,
		down = function(n)
			depth = depth + n
		end,
		up = function(n)
			depth = depth - n
		end,
	}
	for _, v in pairs(input) do
		local opcode = Utility.split(v, " ", true)
		instruction[opcode[1]](tonumber(opcode[2]))
	end
	return depth * horizontalPosition
end

function code:part2(input)
	local input = Utility.split(input, "\n", true)

	local horizontalPosition = 0
	local depth = 0
	local aim = 0

	local instruction = {
		forward = function(n)
			horizontalPosition = horizontalPosition + n
			depth = depth + (aim * n)
		end,
		down = function(n)
			aim = aim + n
		end,
		up = function(n)
			aim = aim - n
		end,
	}

	for _, v in pairs(input) do
		local opcode = Utility.split(v, " ", true)
		instruction[opcode[1]](tonumber(opcode[2]))
	end

	return depth * horizontalPosition
end

return code
